network/netcore/src/multiplexing/yamux.rs

```diff
@@ -76,152 +67,127 @@ where
-        Self {
-            inner: Compat01As03::new(connection),
-        }
+    async fn close(&mut self) -> io::Result<()> {
+        self.inner
+            .close()
+            .await
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))
     }
 }
 
-impl<TSocket> Stream for Listener<TSocket>
+#[async_trait]
+impl<TSocket> StreamMultiplexer for Yamux<TSocket>
 where
-    TSocket: AsyncRead + AsyncWrite + Unpin,
+    TSocket: AsyncRead + AsyncWrite + Send + Debug + Unpin + Sync + 'static,
 {
-    type Item = io::Result<Compat01As03<yamux::StreamHandle<Compat<TSocket>>>>;
-
-    fn poll_next(mut self: Pin<&mut Self>, context: &mut Context) -> Poll<Option<Self::Item>> {
-        match Pin::new(&mut self.inner).poll_next(context) {
-            Poll::Pending => Poll::Pending,
-            Poll::Ready(None) => Poll::Ready(None),
-            Poll::Ready(Some(Err(e))) => {
-                Poll::Ready(Some(Err(io::Error::new(io::ErrorKind::Other, e))))
+    type Substream = StreamHandle;
+    type Control = YamuxControl;
+    type Listener = BoxStream<'static, io::Result<StreamHandle>>;
+
+    async fn start(self) -> (Self::Listener, Self::Control) {
+        let control = self.connection.control();
+        let mut connection = self.connection;
+        let (mut tx, rx) = mpsc::unbounded();
+        // Our library wraps calls to `next_stream` in a future and runs the future on this static
+        // tokio runtime. That way, users of our yamux library don't need to worry about the
+        // unintuitive need for polling for new inbound substreams even if they only need to open
+        // outbound substreams or perform IO on existing susbstreams.
+        let f = async move {
+            loop {
+                let maybe_substream = connection.next_stream().await.transpose();
+                match maybe_substream {
+                    None | Some(Err(_)) => {
+                        // Ignore error.
+                        let _ = tx
+                            .send(Err(io::Error::new(
+                                io::ErrorKind::BrokenPipe,
+                                "Connection closed",
+                            ))
+                                as io::Result<Self::Substream>)
+                            .await;
+                        return;
+                    }
+                    Some(Ok(substream)) => {
+                        if tx.send(Ok(substream)).await.is_err() {
+                            // Failed to notify about new inbound substream.
+                            return;
+                        }
+                    }
+                }
             }
-            Poll::Ready(Some(Ok(substream))) => Poll::Ready(Some(Ok(Compat01As03::new(substream)))),
-        }
+        };
+        // Run the IO future on the executor which drives this connection.
+        tokio::spawn(f);
+        (rx.boxed(), YamuxControl { inner: control })
```

Comment:
i'm still pretty conflicted about this whole block, but it seems like the best compromise for now.

Reply:
any specific concerns? I can at least add that as a comment or `TODO`.